pilot/pkg/networking/apigen/apigen.go

```diff
@@ -0,0 +1,160 @@
+// Copyright 2017 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package apigen
+
+import (
+	"strings"
+
+	gogotypes "github.com/gogo/protobuf/types"
+	golangany "github.com/golang/protobuf/ptypes/any"
+
+	mcp "istio.io/api/mcp/v1alpha1"
+	"istio.io/istio/pilot/pkg/model"
+	"istio.io/istio/pilot/pkg/serviceregistry/external"
+	"istio.io/istio/pkg/config/schema/collections"
+	"istio.io/istio/pkg/config/schema/resource"
+	"istio.io/pkg/log"
+)
+
+// APIGenerator supports generation of high-level API resources, similar with the MCP
+// protocol. This is a replacement for MCP, using XDS (and in future UDPA) as a transport.
+// Based on lessons from MCP, the protocol allows incremental updates by
+// default, using the same mechanism that EDS is using, i.e. sending only changed resources
+// in a push. Incremental deletes are sent as a resource with empty body.
+//
+// Example: networking.istio.io/v1alpha3/VirtualService
```

Comment:
Why do we need this ? isn't GRPC supporting envoy XDS ?

Reply:
Sorry for mixing both apigen and grpcgen - I can split it again if it will make it less confusing, but there are 1 implementation and 1 test file for each.

Apigen is MCP (but using XDS as transport so we don't need 100 files but 1).

grpcgen is GRPC support - a different variant of XDS.